src/Foundation/NSUrlSessionHandler.cs

```diff
@@ -215,13 +215,18 @@ InflightData GetInflightData (NSUrlSessionTask task)
 					if (sessionHandler.inflightRequests.TryGetValue (task, out inflight))
 						return inflight;
 
+				// if we did not manage to get the inflight data, we either got an error or have been canceled, lets cancel the task, that will execute DidCompleteWithError
+				task.Cancel ();
 				return null;
 			}
 
 			public override void DidReceiveResponse (NSUrlSession session, NSUrlSessionDataTask dataTask, NSUrlResponse response, Action<NSUrlSessionResponseDisposition> completionHandler)
 			{
 				var inflight = GetInflightData (dataTask);
 
+				if (inflight == null)
+					return;
+
 				try {
 					// ensure that we did not cancel the request, if we did, do cancel the task
 					if (inflight.CancellationToken.IsCancellationRequested) {
```

Comment:
Ok, I agree that we should move the check of the cancelation token in the inflight data to the GetInflightData method, makes sense, since if we do have the cancelation, we can cancel de NSUrlSessionTask.

This posses an interesting dilema, we can have data from the DidReceiveResponse or DidReceiveData in the inflight data EVEN when the cancelation was requested. In that case, what should we do? We have two options:

1. Get the inflight data, check the cancelation token, cancel if needed and return the data for the handler to deal with. This means that we could be adding headers or body data in the canceled request, but is data we already have and I see no real reason to not do it.
2. Get the inflight data, cancel the task AND return null, instead of the data. In this case, we might be loosing the headers and some body data because the inflight data is cleaned and the task canceled.

I'm more inclined to the first option. For example, we got the headers, and canceled, the user will get the headers and will get the issue in the async read of the stream of the response. 

In summary we are choosing between:

```
     InflightData GetInflightData (NSUrlSessionTask task)
     {
 	var inflight = default (InflightData);
 
 	lock (sessionHandler.inflightRequestsLock)
  		if (sessionHandler.inflightRequests.TryGetValue (task, out inflight)) {
                        if (inflight.CancellationToken.IsCancellationRequested)
			    task?.Cancel ();
  			return inflight;
                }
        task?.Cancel ();
  	return null;
    }
```

And the second one:

```
     InflightData GetInflightData (NSUrlSessionTask task)
     {
 	var inflight = default (InflightData);
 
 	lock (sessionHandler.inflightRequestsLock)
  		if (sessionHandler.inflightRequests.TryGetValue (task, out inflight)) {
                        if (inflight.CancellationToken.IsCancellationRequested) {
			    task?.Cancel ();
                            inflight = null;
                        }
                        return inflight;
                }
        task?.Cancel ();
  	return null;
    }
```